FederatedApi.Characters.api fails with NPE

This failure in `example.federation.FederatedApp` is masked by the race between the effects.
It can be reproduced by replacing line 60 with `service1.exitCode`.

```shell
Fiber failed.
An unchecked error was produced.
java.lang.ExceptionInInitializerError
	at example.federation.FederatedApp$.$anonfun$service1$1(FederatedApp.scala:24)
	at zio.ZManaged.$anonfun$use$4(ZManaged.scala:1089)
	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:925)
	at zio.internal.FiberContext.$anonfun$fork$17(FiberContext.scala:781)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
	at caliban.schema.GenericSchema$$anon$6.toType(Schema.scala:274)
	at caliban.schema.Schema.caliban$schema$Schema$$asType(Schema.scala:39)
	at caliban.schema.Schema.caliban$schema$Schema$$asType$(Schema.scala:39)
	at caliban.schema.GenericSchema$$anon$6.caliban$schema$Schema$$asType$lzycompute(Schema.scala:272)
	at caliban.schema.GenericSchema$$anon$6.caliban$schema$Schema$$asType(Schema.scala:272)
	at caliban.schema.Schema.toType_(Schema.scala:51)
	at caliban.schema.Schema.toType_$(Schema.scala:50)
	at caliban.schema.GenericSchema$$anon$6.toType_(Schema.scala:272)
	at caliban.schema.GenericSchema$$anon$16.toType(Schema.scala:442)
	at caliban.schema.Schema.caliban$schema$Schema$$asType(Schema.scala:39)
	at caliban.schema.Schema.caliban$schema$Schema$$asType$(Schema.scala:39)
	at caliban.schema.GenericSchema$$anon$16.caliban$schema$Schema$$asType$lzycompute(Schema.scala:440)
	at caliban.schema.GenericSchema$$anon$16.caliban$schema$Schema$$asType(Schema.scala:440)
	at caliban.schema.Schema.toType_(Schema.scala:51)
	at caliban.schema.Schema.toType_$(Schema.scala:50)
	at caliban.schema.GenericSchema$$anon$16.toType_(Schema.scala:440)
	at caliban.schema.SchemaDerivation$$anon$1.$anonfun$toType$9(SchemaDerivation.scala:75)
	at caliban.schema.Types$.$anonfun$collectTypes$21(Types.scala:123)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
	at caliban.schema.Types$.collectTypes(Types.scala:122)
	at caliban.schema.Types$.$anonfun$collectTypes$23(Types.scala:124)
	at scala.Option.fold(Option.scala:251)
	at caliban.schema.Types$.$anonfun$collectTypes$21(Types.scala:124)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
	at caliban.schema.Types$.collectTypes(Types.scala:122)
	at caliban.federation.Federation.$anonfun$federate$5(Federation.scala:122)
	at scala.collection.immutable.List.flatMap(List.scala:366)
	at caliban.federation.Federation.federate(Federation.scala:122)
	at caliban.federation.Federation.federate$(Federation.scala:81)
	at caliban.federation.package$.federate(package.scala:3)
	at caliban.federation.Federation$$anon$3.apply(Federation.scala:65)
	at caliban.GraphQL.$at$at(GraphQL.scala:164)
	at caliban.GraphQL.$at$at$(GraphQL.scala:163)
	at caliban.GraphQL$$anon$2.$at$at(GraphQL.scala:150)
	at example.federation.FederatedApi$Characters$.<init>(FederatedApi.scala:82)
	at example.federation.FederatedApi$Characters$.<clinit>(FederatedApi.scala)
	... 7 more

Fiber:Id(1640204743533,1) was supposed to continue to:
  a future continuation at zio.ZIO.run(ZIO.scala:1764)
  a future continuation at zio.ZManaged.use(ZManaged.scala:1088)
  a future continuation at zio.ZIO.exitCode(ZIO.scala:626)

Fiber:Id(1640204743533,1) execution trace:
  at zio.ZManaged.use(ZManaged.scala:1089)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:160)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:153)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO$.bracketExit(ZIO.scala:2437)
  at zio.internal.FiberContext.evaluateNow(FiberContext.scala:565)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO.bracket_(ZIO.scala:291)
  at zio.ZIO.run(ZIO.scala:1764)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:153)
  at zio.ZManaged.map(ZManaged.scala:484)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:160)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:153)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO$.bracketExit(ZIO.scala:2437)
  at zio.internal.FiberContext.evaluateNow(FiberContext.scala:565)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO.bracket_(ZIO.scala:291)
  at zio.ZIO.run(ZIO.scala:1764)
  at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:153)
  at zio.ZManaged.mapM(ZManaged.scala:517)
  at zio.ZManaged.memoize(ZManaged.scala:531)
  at zio.ZIO.once(ZIO.scala:1036)
  at zio.ZRef$.make(ZRef.scala:756)
  at zio.ZManaged.memoize(ZManaged.scala:523)
  at zio.Promise$.makeAs(Promise.scala:243)
  at zio.Promise$.make(Promise.scala:237)
  at zio.ZIO$.fiberId(ZIO.scala:2904)
  at zio.ZIO$.descriptor(ZIO.scala:2676)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO.bracket_(ZIO.scala:291)
  at zio.internal.FiberContext.evaluateNow(FiberContext.scala:561)
  at zio.ZIO.provideSome(ZIO.scala:1266)
  at zio.ZManaged.mapM(ZManaged.scala:517)
  at zio.ZManaged$.releaseMap(ZManaged.scala:2381)
  at zio.ZIO$AccessPartiallyApplied$.apply(ZIO.scala:4455)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZIO.bracket_(ZIO.scala:291)
  at zio.internal.FiberContext.evaluateNow(FiberContext.scala:561)
  at zio.ZIO.provideSome(ZIO.scala:1266)
  at zio.ZIO$.effectSuspendTotal(ZIO.scala:2837)
  at zio.ZManaged.use(ZManaged.scala:1089)
  at zio.ZManaged$ReleaseMap$.make(ZManaged.scala:1375)
  at zio.ZRef$.make(ZRef.scala:756)

Fiber:Id(1640204743533,1) was spawned by:

Fiber:Id(1640204743243,0) was supposed to continue to:
  a future continuation at zio.App.main(App.scala:60)
  a future continuation at zio.App.main(App.scala:59)

Fiber:Id(1640204743243,0) ZIO Execution trace: <empty trace>

Fiber:Id(1640204743243,0) was spawned by: <empty trace>

```

--- a/examples/src/main/scala/example/federation/FederatedApi.scala
+++ b/examples/src/main/scala/example/federation/FederatedApi.scala
@@ -1,15 +1,14 @@
 package example.federation
 
-import example.federation.CharacterService.CharacterService
-import example.federation.EpisodeService.EpisodeService
 import caliban.GraphQL.graphQL
-import caliban.federation.{ EntityResolver, federated }
 import caliban.federation.tracing.ApolloFederatedTracing
+import caliban.federation.{ federated, EntityResolver, GQLKey }
 import caliban.schema.Annotations.{ GQLDeprecated, GQLDescription }
 import caliban.schema.{ ArgBuilder, GenericSchema, Schema }
 import caliban.wrappers.Wrappers.{ maxDepth, maxFields, printSlowQueries, timeout }
 import caliban.{ GraphQL, RootResolver }
-
+import example.federation.CharacterService.CharacterService
+import example.federation.EpisodeService.EpisodeService
 import zio.URIO
 import zio.clock.Clock
 import zio.console.Console
@@ -47,39 +46,50 @@ object FederatedApi {
     case class Subscriptions(characterDeleted: ZStream[CharacterService, Nothing, String])
 
     implicit val roleSchema: Schema[Any, Role]                         = Schema.gen
+    implicit val characterSchema: Schema[CharacterService, Character] =
+      // build this one manually because recursion breaks magnolia
+      obj("Character", None, directives = List(GQLKey("name").directive))(implicit ft =>
+        List(
+          field("name")(_.name),
+          field("nicknames")(_.nicknames),
+          field("origin")(_.origin),
+          field("role")(_.role),
+          field("starredIn")(_.starredIn)
+        )
+      )
     implicit lazy val episodeSchema: Schema[CharacterService, Episode] = Schema.gen
-    implicit val characterSchema: Schema[CharacterService, Character]  = Schema.gen
     implicit val characterArgsSchema: Schema[Any, CharacterArgs]       = Schema.gen
     implicit val charactersArgsSchema: Schema[Any, CharactersArgs]     = Schema.gen
     implicit val episodeArgs: Schema[Any, EpisodeArgs]                 = Schema.gen
     implicit val episodeArgBuilder: ArgBuilder[EpisodeArgs]            = ArgBuilder.gen
 
-    val withFederation = federated(        
+    val withFederation = federated(
       EntityResolver.from[CharacterArgs](args => ZQuery.fromEffect(CharacterService.findCharacter(args.name))),
       EntityResolver.from[EpisodeArgs](args =>
-          ZQuery
-            .fromEffect(CharacterService.getCharactersByEpisode(args.season, args.episode))
-            .map(characters =>
-              Some(
-                Episode(
-                  args.season,
-                  args.episode,
-                  ZQuery.succeed(characters)
-                )
+        ZQuery
+          .fromEffect(CharacterService.getCharactersByEpisode(args.season, args.episode))
+          .map(characters =>
+            Some(
+              Episode(
+                args.season,
+                args.episode,
+                ZQuery.succeed(characters)
               )
             )
-        ))
+          )
+      )
+    )
 
     val api: GraphQL[Console with Clock with CharacterService] =
-        graphQL(
-          RootResolver(
-            Queries(
-              args => CharacterService.getCharacters(args.origin),
-              args => CharacterService.findCharacter(args.name)
-            ),
-            Mutations(args => CharacterService.deleteCharacter(args.name))
-          )
-        ) @@ standardWrappers @@ withFederation
+      graphQL(
+        RootResolver(
+          Queries(
+            args => CharacterService.getCharacters(args.origin),
+            args => CharacterService.findCharacter(args.name)
+          ),
+          Mutations(args => CharacterService.deleteCharacter(args.name))
+        )
+      ) @@ standardWrappers @@ withFederation
   }
 
   object Episodes extends GenericSchema[EpisodeService] {
@@ -95,18 +105,18 @@ object FederatedApi {
     implicit val episodeSchema: Schema[Any, Episode]           = Schema.gen
 
     val api: GraphQL[Console with Clock with EpisodeService] =
-        graphQL(
-          RootResolver(
-            Queries(
-              args => EpisodeService.getEpisode(args.season, args.episode),
-              args => EpisodeService.getEpisodes(args.season)
-            )
+      graphQL(
+        RootResolver(
+          Queries(
+            args => EpisodeService.getEpisode(args.season, args.episode),
+            args => EpisodeService.getEpisodes(args.season)
           )
-        ) @@ standardWrappers @@ federated(
+        )
+      ) @@ standardWrappers @@ federated(
         EntityResolver.from[EpisodeArgs](args =>
           ZQuery.fromEffect(EpisodeService.getEpisode(args.season, args.episode))
         )
-        )
+      )
 
   }
 

